google-cloud-spanner/src/main/java/com/google/cloud/spanner/spi/v1/HeaderInterceptor.java

```diff
@@ -139,4 +147,32 @@ private TagContext getTagContext(Metadata headers, String method) {
     }
     return getTagContext(method, projectId, instanceId, databaseId);
   }
+
+  private Attributes getMetricAttributes(Metadata headers, String method) {
```

Comment:
(I see that this is consistent with how `getTagContext` works, so I understand where it is coming from)

But this feels like a relatively expensive operation that we execute for __every__ request that we execute. Both this method and `getTagContext` match a regular expression against one of the headers to get the project, instance, and database, and then build a data structure. That structure is then only used if a number of conditions are met (metric collection has been enabled + server timing key is returned).

Maybe for a separate PR, but I would suggest that we look into:
1. parsing the project, instance, and database only once from the header
2. Only parse that information and build the data structure if we know that they will actually be used (so basically move it into the if statements above)

Reply:
1) IMK, One spanner client can have multiple database object, so we need ti parse Header to get database every time ? 
We can save project and instance though and use it again. 
2) You mean we should call this `getMetricAttributes` in the `processHeader` method, right ?

Reply:
done